### PR TITLE
Initial PR gsoc18 : PopupEventController

### LINKED ISSUE #257: Create class to handle popovers during a scenario
## Description
As a developer,
I need create a controller class to handle presenting and interactions for in-sequence popovers,
so that I can ensure the new content follows efficient model-controller principles, and is sufficiently sandboxed from existing code.

## Mocks
n/a

### Update [Required]
- [x] handles presenting parent view
- [x] handles presenting media from the model class
- [x] handles interactions
## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

### LINKED ISSUE #258: Create a model class to describe popovers
## Description
As a developer,
I need to create a model class to describe popovers,
so that I can ensure structure for creating new content is clear and usable.

## Mocks
n/a

### Update [Required]
- [x] owns references to media
- [x] describes animations
- [x] describes event order and timing
## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

### LINKED ISSUE #259: Add functionality to access popover
## Description
As a developer,
I need to add a new field to the Answer table in the database, and add logic to the scenario view controller,
so that the content can be identified and accessed with minimal changes to existing code.

## Mocks
n/a
### Update [Required]
- [x] Answer table includes new field with neutral data for existing scenario events
- [x] Scenario view controller checks for the need to present a popover
- [x] Scenario view controller creates an instance controller class when necessary
## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

### PR DESCRIPTION
**Pull request by justKD**
### Description
Fixes #257
Fixes #258
Fixes #259

- Fixed storyboard version inconsistencies after updating Xcode
- added but removed work on a visualization tool for scenario maps - it exists in its own repo now
- duplicated sqlite db to create a test version for edits and new content
- added files for new classes to handle out-of-context story events (sequences and popups)
- updated scenario 5 (home) with dummy data for testing ooc events
- PopupEventController is a subclass of UIView and handles displaying the popup, animating the view, timing the event for automatic dismissal, handles tap interaction for manual dismissal, handles creating an instance of AVAudioPlayer (via SoundPlayer class) and playing a sound. Ultimately, it will be a controller for instances of the PopupEventModel class, but currently uses static data. I'm using the view itself as the controller rather than abstracting to another level in order to better make use of ARC. Example described below.
- ScenarioViewController declared PopupEventControllerDelegate, ensures instances of PopupEventController are released from memory
- Added SoundPlayer class to handle instanced sound. Currently used by PopupEventController.

Example gif

### Type of Change:
**Delete irrelevant options.**

- Code
- Quality Assurance
- User Interface

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Describe the tests you ran to verify your changes. Provide instructions or GIFs so we can reproduce. List any relevant details for your test.

I've set up the dummy test data to test each use case. The PopupEventPlayer class is self-contained, only requiring a few lines of code to instantiate and call. There's no need to create a global variable on any class that uses this class.

```
let newPopup : PopupEventPlayer? = PopupEventPlayer(delegate: self)
guard let popup = newPopup else {return}
self.view.addSubview(popup)
```
Declare the calling class as PopupEventPlayerDelegate, and add the delegate method:
```
func popupDidFinish(sender: PopupEventPlayer) {
    sender.removeFromSuperview()
}
```
When the popup is dismissed, this delegate method is called. The instance can then be removedFromSuperview which, since the setup creates a local instance, we know it releases from memory since there are no more strong references.

This makes it easy to test the class as a whole since it is fairly self-contained. I've added the logic to detect if a popup should happen based on the dummy data in the test database, and have run through all of the necessary options to see that the popup is called or not as expected. I've also tested calling multiple popups in quick succession.

### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] Any dependent changes have been merged 

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

### CODE REVIEW

File: Powerup/OOC-Event-Classes/PopupEventPlayer.swift [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    MARK: Properties
+    ******************************* */
+    weak var delegate:PopupEventPlayerDelegate?
+    var soundPlayer : SoundPlayer? = SoundPlayer()
+    
+    var width : CGFloat,
+        height : CGFloat
+    
+    var bgColor : UIColor { didSet {updateContainer()} }
+    var borderColor : UIColor { didSet {updateContainer()} }
```
[the remark this answers is not in the document]
Comment (justKD) [reply Powerup/OOC-Event-Classes/PopupEventPlayer.swift]: I can make the change if anyone thinks it's necessary, but I disagree with the rule. Making the change would add tons of unnecessary white space when the intent and function is perfectly readable here.

File: Powerup/OOC-Event-Classes/SoundPlayer.swift [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+//  Copyright © 2018 Systers. All rights reserved.
+//
+
+import UIKit
+import AVFoundation
+import AudioToolbox
+
+/*
+    Example Use
+
```
[the remark this answers is not in the document]
Comment (justKD) [reply Powerup/OOC-Event-Classes/SoundPlayer.swift]: Same as above. Just a line break. No trailing white space.

File: Powerup/OOC-Event-Classes/PopupEventPlayer.swift [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+import UIKit
+import AVFoundation
+
+class PopupEventPlayer : UIView {
+    /* *******************************
+    MARK: Properties
+    ******************************* */
+    weak var delegate:PopupEventPlayerDelegate?
+    var soundPlayer : SoundPlayer? = SoundPlayer()
+
```
[the remark this answers is not in the document]
Comment (justKD) [reply Powerup/OOC-Event-Classes/PopupEventPlayer.swift]: I'm not sure why this is flagged. I've checked, and there's no more white space here than any other line break. No trailing whitespace.

File: Powerup/OOC-Event-Classes/PopupEventPlayer.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (sunjunkie) [reply Powerup/OOC-Event-Classes/PopupEventPlayer.swift]: I agree. Let's keep the code as is.

File: Powerup/OOC-Event-Classes/PopupEventPlayer.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (sunjunkie) [reply Powerup/OOC-Event-Classes/PopupEventPlayer.swift]: I'm seeing spaces at the beginning of all the blank lines. That may be it.

File: Powerup/OOC-Event-Classes/SoundPlayer.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (sunjunkie) [reply Powerup/OOC-Event-Classes/SoundPlayer.swift]: I'm seeing whitespace at the beginning of the blank lines. That may be it.

[the remark this answers is not in the document]
Comment (justKD) [reply Powerup/OOC-Event-Classes/SoundPlayer.swift]: I'm pretty sure that's just the tabbed indent that Xcode adds to every new line. May mentioned turning off Codacy for all repos until we add yaml files with custom rules. I'm not sure if that's been done here. I know some of the other projects were having much bigger issues with the bot.

screen shot 2018-05-17 at 10 11 11 pm

**Comment by sunjunkie** [comment]
This looks good. 

- I think the Codacy whitespace problem is whitespace on blank lines.
- Yes, ignore the Codacy complaint about the trailing brace
- Use the gsoc18-code branch for your work.
- Take a look at https://developer.apple.com/library/content/documentation/Xcode/Reference/xcode_markup_formatting_ref/index.html#//apple_ref/doc/uid/TP40016497-CH2-SW1 for how to comment/markup Swift code to take advantage of Quick Help and code completion in Xcode.

**Comment by justKD** [comment]
Primarily issue 257. I updated the PR title to reflect that.

**Comment by sunjunkie** [comment]
You can add "Fixes #<issue number>" just under "Description". Then Github will automatically close the issue when the PR is merged.

**Comment by sunjunkie** [comment]
Don't add the fixes thing then til the issue is fixed. For the future, I would break the issues down further if you think the PR's will get too big so we can link a PR to an issue.

**Comment by justKD** [comment]
I've pushed another commit and this PR has been updated and addresses the whole issue.

**Comment by sunjunkie** [comment]
Scenario 5 shows the mine sweeper game after the scenario is complete. Any reason for that?
Is scenario 5 your test scenario instead of scenario 1?

**Comment by justKD** [comment]
The "home" scenario is the first scenario in the game, and it was already numbered as 5. In the database they are out of order. Scenario 1 is the school, and scenario 2 has entries in the DB but doesn't exist in the game.

**Comment by justKD** [comment]
@SIMRAN1 Thanks! Swimat is great!

**Comment by sunjunkie** [comment]
@justKD Nice work!

**Pull request merged**